File: 2015/Day 01/Go/main.go
package main

import (
	"fmt"
	"os"
	"log"
)

func main() {
	content, err := os.ReadFile("input.txt")
	if err != nil {
			log.Fatal(err)
	}

	var floor uint8 = final_floor(content)
	var basement_position = basement(content)

	fmt.Printf("Final floor: %d\nFirst basement position: %d", floor, basement_position)
}

func final_floor(content_string []byte)uint8{
	var floor uint8 = 0

	for i:=0;i<len(content_string);i++ {
		if content_string[i] == '('{
			floor++
		} else if content_string[i] == ')' {
			floor--
		}
	}
	return floor
}

func basement(content_string []byte)int{
	var floor int8 = 0

	for i:=0;i<len(content_string);i++ {
		if content_string[i] == '('{
			floor++
		} else if content_string[i] == ')' {
			floor--
		}
		if floor == -1{
			return i+1
		}
	}
	return -1
}

File: 2015/Day 01/C/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "fileio.h"

int final_floor(char *filecon);
int basement(char *filecon);

int main(void) { 
    char *filecon = (char*)malloc(MAX_DATA_SIZE*sizeof(char));
    if (filecon == NULL) {
        perror("Error allocating memory");
        return 1;
    }
    int status = open_file("input.txt", filecon);
    if (status != 0) {
        return status;
    }
    int floor_ans = final_floor(filecon);
    int basement_ans = basement(filecon);
    printf("Final floor: %d\nFirst basement position: %d", floor_ans, basement_ans);
    free(filecon);
    return 0; 
}

int final_floor(char *filecon) {
  int i;
  int floor_num = 0;
  for(i=0; i<strlen(filecon); i++) {
      if (filecon[i] == '(') {
          floor_num++;
      } else if (filecon[i] == ')') {
          floor_num--;
      }
  }
  return floor_num;
}

int basement(char *filecon) {
  int i;
  int floor_num = 0;
  for(i=0; i<strlen(filecon); i++) {
      if (filecon[i] == '(') {
          floor_num++;
      } else if (filecon[i] == ')') {
          floor_num--;
      }
      if (floor_num == -1) {
          return i+1;
      }
  }
  return -1;
}
